src/buffer.c

```diff
@@ -103,11 +103,14 @@ int buffer_meth_send(lua_State *L, p_buffer buf) {
 * object:receive() interface
 \*-------------------------------------------------------------------------*/
 int buffer_meth_receive(lua_State *L, p_buffer buf) {
-    int err = IO_DONE, top = lua_gettop(L);
+    int err = IO_DONE, top;
     luaL_Buffer b;
     size_t size;
     const char *part = luaL_optlstring(L, 3, "", &size);
     timeout_markstart(buf->tm);
+    /* make sure we don't confuse buffer stuff with arguments */
+    lua_settop(L, 3);
+    top = lua_gettop(L);
```

Comment:
Isn't this line redundant and always equivalent to `top = 3` here? I'm not actually sure what the point of the `top` variable is, but this would change the behavior.

Reply:
The behavior in the patch is different from `top=3`, as `lua_settop()` will extend the stack to make sure that subsequent buffer allocations don't clobber optional parameters. The way this is done is consistent with the original patch from Diego in f329aae7 (which just missed this particular instance).